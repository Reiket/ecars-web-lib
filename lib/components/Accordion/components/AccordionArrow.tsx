import type {FC} from 'react';
import {Icons} from '@/services/icons';

export const AccordionArrow: FC = () => <Icons.ArrowNarrowDown className="accordion__arrow" />;
