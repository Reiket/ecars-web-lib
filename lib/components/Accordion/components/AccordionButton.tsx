import type {FC, ReactNode} from 'react';

interface Props {
  children: ReactNode;
  onClick: () => void;
}

export const AccordionButton: FC<Props> = ({onClick, children}) => (
  <button
    onClick={onClick}
    type="button"
    className="accordion__button"
  >
    {children}
  </button>
);
