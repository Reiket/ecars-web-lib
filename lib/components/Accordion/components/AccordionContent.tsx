import type {FC} from 'react';
import {useRef} from 'react';

export interface Props {
  isOpen: boolean;
  children: string;
}

export const AccordionContent: FC<Props> = ({isOpen, children}) => {
  const contentHeight = useRef<HTMLParagraphElement | null>(null);
  const accordionStyle = isOpen ? {height: contentHeight.current?.scrollHeight} : {height: 0};
  return (
    <div
      style={accordionStyle}
      className="accordion__content"
    >
      <p
        className="accordion__text"
        ref={contentHeight}
      >
        {children}
      </p>
    </div>
  );
};
