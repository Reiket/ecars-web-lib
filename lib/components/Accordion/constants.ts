export const ACCORDION_TEST_ID = 'accordionTestId';
