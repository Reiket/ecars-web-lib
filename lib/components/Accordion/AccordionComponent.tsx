import type {FC} from 'react';
import {Accordion} from '@/components/Accordion/constants';
import type {ElementProps} from '@/services/types';

export interface Props extends ElementProps {
  isOpen: boolean;
  textContent: string;
  title: string;
  onClick: () => void;
}

export const AccordionComponent: FC<Props> = ({block, isOpen, textContent, title, onClick}) => (
  <Accordion.Layout
    isOpen={isOpen}
    block={block}
  >
    <Accordion.Block
      title={title}
      onClick={onClick}
    />
    <Accordion.Content isOpen={isOpen}>{textContent}</Accordion.Content>
  </Accordion.Layout>
);
