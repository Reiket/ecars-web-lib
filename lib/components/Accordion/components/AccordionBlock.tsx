import type {FC} from 'react';
import {Accordion} from '../constants';
import type {AccordionTitleProps} from '@/components/Accordion/components/AccordionTitle';

export interface Props extends AccordionTitleProps {
  onClick: () => void;
}

export const AccordionBlock: FC<Props> = ({onClick, title}) => (
  <Accordion.Button onClick={onClick}>
    <Accordion.Title title={title} />
    <Accordion.Arrow />
  </Accordion.Button>
);
