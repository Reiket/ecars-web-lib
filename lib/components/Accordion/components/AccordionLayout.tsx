import type {FC, ReactNode} from 'react';
import {cn} from '@/services/helpers';
import type {ElementProps} from '@/services/types';
import {ACCORDION_TEST_ID} from '@/components/Accordion/constants';

export interface Props extends ElementProps {
  children: ReactNode;
  isOpen: boolean;
}

export const AccordionLayout: FC<Props> = ({block, isOpen, children}) => (
  <div
    data-testid={ACCORDION_TEST_ID}
    className={cn(block, 'accordion', {
      '_active': isOpen,
    })}
  >
    {children}
  </div>
);
