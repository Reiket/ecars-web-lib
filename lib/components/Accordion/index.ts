import {AccordionComponent} from '@/components/Accordion/AccordionComponent';
import {AccordionContent} from '@/components/Accordion/components/AccordionContent';
import {AccordionArrow} from '@/components/Accordion/components/AccordionArrow';
import {AccordionTitle} from '@/components/Accordion/components/AccordionTitle';
import {AccordionBlock} from '@/components/Accordion/components/AccordionBlock';
import {AccordionLayout} from '@/components/Accordion/components/AccordionLayout';
import {AccordionButton} from '@/components/Accordion/components/AccordionButton';

export const Accordion = Object.assign(AccordionComponent, {
  Content: AccordionContent,
  Arrow: AccordionArrow,
  Title: AccordionTitle,
  Block: AccordionBlock,
  Layout: AccordionLayout,
  Button: AccordionButton,
});
