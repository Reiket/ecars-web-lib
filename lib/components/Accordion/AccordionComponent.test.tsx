import {fireEvent, render, screen} from '@testing-library/react';
import {Accordion, ACCORDION_TEST_ID} from '@/components/Accordion/constants';
import {AccordionExample} from '@src/components/AccordionExample';
import {accordionContentMock} from '@src/services/mocks';

const defaultProps = {
  isOpen: false,
  textContent: 'Test content',
  title: 'Test title',
  onClick: jest.fn(),
};
describe('Accordion Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const isOpenValues = [true, false];
  const renderAccordion = (props = {}) => {
    const mergedProps = {...defaultProps, ...props};
    return render(<Accordion {...mergedProps} />);
  };

  test('renders the accordion component and title, text correctly', () => {
    const {container} = renderAccordion();
    expect(screen.getByText(defaultProps.title)).toBeInTheDocument();
    expect(screen.getByText(defaultProps.textContent)).toBeInTheDocument();
    expect(screen.getByTestId(ACCORDION_TEST_ID)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  isOpenValues.forEach((isOpen) => {
    const isOpenString = `should ${isOpen ? 'open' : 'close'} the content`;
    test(isOpenString, () => {
      const {container} = renderAccordion({isOpen});
      const content = screen.getByText(defaultProps.textContent);

      if (isOpen) {
        expect(content).toBeVisible();
      } else {
        expect(content.offsetHeight).toBe(0);
        expect(container).toMatchSnapshot();
      }
    });
  });
  test('triggers onClick handler when title is clicked', () => {
    renderAccordion();
    fireEvent.click(screen.getByText(defaultProps.title));
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
  });
  test('closes previously opened accordion when another is clicked', () => {
    render(<AccordionExample items={accordionContentMock} />);

    const getAccordionElements = () =>
      accordionContentMock.map((item) => ({
        titleElement: screen.getByText(item.title),
        contentElement: screen.getByText(item.text),
      }));

    const accordionElements = getAccordionElements();

    const checkAccordionVisibility = (clickIndex: number) => {
      accordionElements.forEach((_, index) => {
        const isVisible = index === clickIndex;
        const content = accordionElements[index].contentElement;

        if (isVisible) {
          expect(content).toBeVisible();
        } else {
          expect(content.offsetHeight).toBe(0);
        }
      });
    };

    accordionElements.forEach((_, clickIndex) => {
      fireEvent.click(accordionElements[clickIndex].titleElement);
      checkAccordionVisibility(clickIndex);
    });
  });
});
