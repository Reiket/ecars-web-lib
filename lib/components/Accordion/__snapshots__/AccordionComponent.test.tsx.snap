// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component renders the accordion component and title, text correctly 1`] = `
<div>
  <div
    class="accordion"
    data-testid="accordionTestId"
  >
    <button
      class="accordion__button"
      type="button"
    >
      <h3
        class="accordion__title"
      >
        Test title
      </h3>
      <svg
        class="accordion__arrow"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
    </button>
    <div
      class="accordion__content"
      style="height: 0px;"
    >
      <p
        class="accordion__text"
      >
        Test content
      </p>
    </div>
  </div>
</div>
`;

exports[`Accordion Component should close the content 1`] = `
<div>
  <div
    class="accordion"
    data-testid="accordionTestId"
  >
    <button
      class="accordion__button"
      type="button"
    >
      <h3
        class="accordion__title"
      >
        Test title
      </h3>
      <svg
        class="accordion__arrow"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
    </button>
    <div
      class="accordion__content"
      style="height: 0px;"
    >
      <p
        class="accordion__text"
      >
        Test content
      </p>
    </div>
  </div>
</div>
`;
